lms/services/organization.py

```diff
@@ -146,8 +169,73 @@ def update_organization(self, organization, name=None, enabled=None, notes=None)
         if notes is not None:
             organization.settings.set("hypothesis", "notes", notes)
 
+        # Use the fact the default is ... to allow us to blank or set based on
+        # the param being present
+        if parent_public_id is None:
+            organization.parent_id = None
+        elif parent_public_id is not ...:
+            self._move_organization_parent(organization, parent_public_id)
+
         return organization
 
+    def _move_organization_parent(self, organization: Organization, parent_public_id):
+        """Change an organizations parent, without creating loops"""
+
+        # This would be caught by the next check, but doing it here means we
+        # can give a more sensible error message
+        if parent_public_id == organization.public_id:
+            raise InvalidOrganizationParent(
+                "Cannot set an organization to be it's own parent"
+            )
+
+        # Get a list including our self and all are children etc.
+        for item in self._get_hierarchy(organization.id):
+            if item.public_id == parent_public_id:
+                raise InvalidOrganizationParent(
+                    f"Cannot use '{parent_public_id}' as a parent as it a "
+                    "child of this organization"
+                )
+
+        parent = self._organization_search_query(
+            public_id=parent_public_id
+        ).one_or_none()
+        if not parent:
+            raise InvalidOrganizationParent(
+                f"Could not find parent organization: '{parent_public_id}'"
+            )
+
+        organization.parent = parent
+
+    def _get_hierarchy(self, id_) -> List[Organization]:
+        """Get an organization and all it's children (order is not guaranteed).
+
+        :param id_: Organization id to look for
+        """
+
+        # Find all the relevant orgs in the hierarchy by id using a recursive
+        # CTE
+        cols = [Organization.id, Organization.parent_id]
+        base_case = (
+            self._db_session.query(*cols).filter(Organization.id == id_)
+            # The name of the CTE is arbitrary, but must be present
+            .cte("organizations", recursive=True)
+        )
+
+        # We are going to self join onto the above with anything that is a
+        # child of the objects we've only seen
+        join_condition = Organization.parent_id == base_case.c.id
+        recursive_case = self._db_session.query(*cols).join(base_case, join_condition)
+
+        # This will recurse until no new rows are added
+        rows = self._db_session.query(base_case.union(recursive_case)).all()
```

Comment:
It might be the type of comment you get asked to remove (by me?) in another review but a

```
this generates a query like:

WITH RECURSIVE organizations(id, parent_id) AS 
 (SELECT organization.id AS id, organization.parent_id AS parent_id 
 FROM organization 
 WHERE organization.id = %(id_1)s 

UNION SELECT organization.id AS organization_id, organization.parent_id AS organization_parent_id 
 FROM organization JOIN organizations ON organization.parent_id = organizations.id OR organization.id = organizations.parent_id)
  SELECT organizations.id AS organizations_id, organizations.parent_id AS organizations_parent_id 
 FROM organizations
```

or maybe that will be just noise?

Reply:
I might point to an article about recursive queries in general, but I'm a bit allergic to queries which need a lot of maintenance to be correct. If the query changed, and we didn't update it, I'm worried it would be misleading.